writeaheadlog.go

```diff
@@ -156,17 +155,61 @@ func newWal(path string, deps dependencies) (u []Update, w *WAL, err error) {
 
 // readWALMetadata reads WAL metadata from the input file, returning an error
 // if the result is unexpected.
-func readWALMetadata(data []byte) error {
-	var md metadata
-	dec := json.NewDecoder(bytes.NewReader(data))
-	if err := dec.Decode(&md); err != nil {
-		return build.ExtendErr("error reading WAL metadata", err)
-	} else if md.Header != metadataHeader {
-		return errors.New("WAL metadata header does not match header found in WAL file")
-	} else if md.Version != metadataVersion {
-		return errors.New("WAL metadata version does not match version found in WAL file")
+func readWALMetadata(data []byte) (uint16, error) {
+	buffer := bytes.NewBuffer(data)
+
+	// Data should at least be long enough to contain the recoveryState,
+	// headerLength and versionLength
```

Comment:
Don't have to read lengths if  you know they are a fixed size. Since we don't need the rest of the first page for anything else, my vote would just be to make the name 16 bytes (a string), the version 16 bytes (a string), and then the recovery state 8 bytes (a uint64). Then you can read into byte arrays and don't need to do any complex length management.

Reply:
I thought it would make sense. This way the metadata can be changed in the future without worrying about the length. It doesn't really make reading/writing it slower and the lengths only take up 4 bytes.

Reply:
Yes but for example you forgot to do bounds checks on the size of the header when reading into disk. If for some reason (wrong file perhaps) the header bytes were super large, you'd call `make` on a slice that's too big, and you'd probably crash the program.

It simplifies the code and the edge cases a lot to just have a constant sized header. We do this a few places in the sia repo using a `types.Specifier`.

Reply:
Let's punt this to the next PR though